### ISSUE TITLE
Docker 1.13 cli gives a warning about crypto/x509

### ISSUE THREAD

**Issue by JettJones** [docker/docker#30450]
(moving from https://github.com/docker/for-win/issues/422)

### Expected behavior
Docker ouput should be free of warnings (unless there's something to fix)

### Actual behavior
Docker 1.13.0 provides a developer warning on every run - it is reporting a library does not exist on windows. That's not something I can fix as a user.
```
>docker --version
time="2017-01-24T09:45:40-08:00" level=warning msg="Unable to use system certificate pool: crypto/x509: system root pool is not available on Windows"
Docker version 1.13.0, build 49bf474
```
### Information
  - host distribution and version (Windows version, build number, etc)
running on windows 10 pro, Version 1607 (OS Build 14292.693)

This might be an issue from Go's system cert pool: https://github.com/golang/go/issues/16736
This may also be fixed if https://github.com/docker/docker/issues/24806 is completed.

### Steps to reproduce the behavior

  1. Install docker 1.13.0 on windows
  2. Run `docker --version` (or any other command)

**Comment by denialg** [comment]
Same issue on Windows 7 Enterprise 64-bit

**Comment by thaJeztah** [comment]
ping @dmcgowan @jhowardmsft @johnstep ptal

**Comment by dmcgowan** [comment]
That is pretty obnoxious, if it is not actionable we should probably have it at INFO level. It will be fixed once we are building in 1.8 but considering it is non-critical we don't need it as a warning. Also curious why it is getting executed on every invocation rather than only when certificates are used.

**Comment by stevvooe** [comment]
@dmcgowan Warning is the right level here.

What is more disconcerting is why we are initializing the certs for the `version` command. Shouldn't these only get loaded if we create client?

**Comment by aaronlehmann** [comment]
`version` connects to the server to get the server version.

**Comment by cpuguy83** [comment]
`--version` does not need to contact the server since it should only be showing the client version, but it is apparently.

**Comment by aaronlehmann** [comment]
Ah, saw "`version` command" above and thought this was about `docker version`.

**Comment by xtrycatchx** [comment]
i encountered the same error. im using latest docker tool box

> $ docker --version
> time="2017-02-03T14:59:50+08:00" level=warning msg="Unable to use system certifi
> cate pool: crypto/x509: system root pool is not available on Windows"
> Docker version 1.13.0, build 49bf474

**Comment by galessandro** [comment]
I also have the same error. I recently download the latest doocker tool box. I am behind a proxy corporation

 ~ $ docker --version
time="2017-02-03T11:54:10-05:00" level=warning msg="Unable to use system certifi  cate pool: crypto/x509: system root pool is not available on Windows"
Docker version 1.13.0, build 49bf474

Any clue to fix that?

**Comment by dmcgowan** [comment]
@stevvooe seems to be part of the initialization of the client.

See https://github.com/docker/docker/blob/master/cmd/docker/docker.go#L48

Initialize creates an api client, which creates an http client, which uses the tls config. Only running the setup on commands which need the client would be tricky, but I don't think we need to load custom certs from the client anyway unless Docker Content Trust is used. It would require some refactoring to separate this out, in the meantime, I think the warning is unnecessary since the custom certs are not relevant to user and nothing can be done. In Go 1.8 this will all be irrelevant.

**Comment by stevvooe** [comment]
@dmcgowan After talking with @cyli, I think that the problem with that code is that it returns an empty pool for windows when it should actually return a nil pool. On windows, cert lookup falls back to a function and having an empty pool means we have no certs in place. Please confirm, but that should fix.

Good thing this warning was in place or no one would now that this is broken at all.

**Comment by cyli** [comment]
@stevvooe After confirming with @dmcgowan looks like in docker itself, a nil pool  if a custom cert is found:  https://github.com/docker/docker/blob/master/registry/registry.go#L64.  If no custom certs are found, a `nil` pool is used.

**Comment by toedter** [comment]
Is this planned to be addressed in 1.13.1?
In 1.13.1-rc2 the warning is still there.

**Comment by cyli** [comment]
@toedter It should have made it into 1.13.1-rc2, as per the release notes on https://github.com/docker/docker/releases/tag/v1.13.1-rc2 - it's being logged as info now, rather than a warning.

**Comment by toedter** [comment]
@cyli yes, thx.
Also in the 1.13.1 release it is logged as info but this is still a bit annoying since it is logged with every docker command, e.g.

> docker --version
time="2017-02-08T22:29:39+01:00" level=info msg="Unable to use system certificate pool: crypto/x509: system root pool is not available on Windows"
Docker version 1.13.1, build 092cba3

Since this is nothing a Windows user can do about (as far as I know), it would be great if this info log could be suppressed somehow.

**Comment by cyli** [comment]
@toedter Just double-checking - do you have custom CA certs configured for a private registry, etc?  IIRC, if you don't, it should not emit that warning - which was left in because one unfortunately has the choice of either using custom CA certs or trusting the system store.  If someone configures the custom certs, we wanted to warn users that the docker client would be unable to use the system certificate.

**Comment by nerro** [comment]
@cyli,
I confirm, that the issue exists (I have no custom CA certs configured, etc.)
```
$ docker version
time="2017-02-09T04:21:17+01:00" level=info msg="Unable to use system certificate pool: crypto/x509: system root pool is not available on Windows"
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 08:47:51 2017
 OS/Arch:      windows/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 08:47:51 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Comment by cyli** [comment]
@nerro Thank you - it definitely shouldn't be happening in the no CA cert case.
cc @dmcgowan

**Comment by kdh6429** [comment]
Is this fixed? still on window 10 Pro.

>docker-machine version
docker-machine version 0.9.0, build 15fd4c7

>docker --version
time="2017-02-10T11:39:45+09:00" level=warning msg="Unable to use system certificate pool: crypto/x509: system root pool is not available on Windows"
Docker version 1.13.0, build 49bf474

**Comment by cyli** [comment]
Oddly I'm not seeing the same issue on Windows Server 2016.  Will try to find a Windows 10 box to try.

**Comment by junxy** [comment]
same issue
```
$ docker version
time="2017-02-10T11:43:21+08:00" level=info msg="Unable to use system certificate pool: crypto/x509: system root pool is not available on Windows"
Client:
 Version:      1.13.1
 API version:  1.24 (downgraded from 1.26)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 08:47:51 2017
 OS/Arch:      windows/amd64

Server:
 Version:      1.12.2
 API version:  1.24 (minimum version )
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 17:00:50 2016
 OS/Arch:      linux/amd64
 Experimental: false
```

**Comment by cyli** [comment]
@dmcgowan Found it - it's not coming from `ReadCertsDirectory`, which i think would only be called from the daemon or from trust commands on the client, which no one here seems to be using.

It's coming from https://github.com/docker/docker/blob/master/cli/command/cli.go#L246, when they're using `docker-machine`.  The remote host's CA cert is provided as an environment variable.

I'd argue that against a remote daemon, we probably don't even want the union of the system certs and custom certs - we'd either want an exclusive custom pool, or the system pool if no custom certs are provided.  Maybe that can be an option in `go-connections`?

**Comment by thaJeztah** [comment]
ping @nathanleclaire

**Comment by matheuscmpm** [comment]
Any ideas about how to fix/work around this on Windows? My dockerfile can be flawless executed on Mac and Linux, but not on Windows. I guess something related to this issue does not make possible to mount my volumes inside of my VM and it breaks my http container.

**Comment by nathanleclaire** [comment]
> I'd argue that against a remote daemon, we probably don't even want the union of the system certs and custom certs - we'd either want an exclusive custom pool, or the system pool if no custom certs are provided. Maybe that can be an option in go-connections?

This seems like the right solution to me.

Although I'm a bit confused why the Docker client reports that the "system root pool is not available on Windows". Why isn't it? Does this mean that certificates signed by a "proper" 3rd party CA won't work?

**Comment by cyli** [comment]
@matheuscmpm The message only appears in windows to indicate that we cannot validate the union of custom + system certs, and can do only one or the other.  So if you provide us with a custom self-signed CA cert to validate against (as, for instance, docker machine does or content trust may against custom trust servers), then we will ONLY be able to validate against that custom cert, and not the custom cert + the system certs.  Or you could not provide a custom cert, and it'd successfully validate against the system pool.

This is probably not related to mounting volumes inside the VM, since if there were an issue validating the CA cert at all, it would also affect the ability to create containers or do anything else inside the VM, and not just mount containers.

**Comment by cyli** [comment]
@nathanleclaire:  That comes from trying to call `x509.SystemCertPool()` - which is not supported in go 1.7 and I think go 1.8 as well AFAIK.  So you can't get a pool with the system certs that you can add additional certs to.  However, if you [do not provide a root pool to the TLS verify opts](https://github.com/golang/go/blob/release-branch.go1.7/src/crypto/x509/verify.go#L239), it will [use the windows CryptoAPI calls](https://github.com/golang/go/blob/release-branch.go1.7/src/crypto/x509/root_windows.go#L143) to verify against the system roots.

**Comment by matheuscmpm** [comment]
@cyli I see. That's weird, I am using my folders inside of c:/Users/username/ to mount my volumes. The exact same dockerfile/project can be executed by MacOS or Ubuntu, without any problems/errors, so I thought this was causing my problem. I am out of ideas, do you know where I can find this kind of solution to my problem?

**Comment by cyli** [comment]
@matheuscmpm Just to verify that I understand the issue: the container gets created successfully, just without the relevant local data?  Or does the container not even start?

Also, apologies I'm not very familiar with docker on windows, but possibly some relevant issues to check out are the ones here:  https://github.com/docker/for-win/issues.  I also found this, which I'm not sure is related:  https://github.com/docker/docker/issues/30355.

If none of those fit, it might be worth filing a new issue with more details about your setup, though.

**Comment by nathanleclaire** [comment]
If using Machine, the `C:\Users` directory is mounted into `/c/users` in the boot2docker VM, sometimes there are path parsing issues with Compose or otherwise, you may want to take a look at `COMPOSE_CONVERT_WINDOWS_PATHS`, https://docs.docker.com/compose/reference/envvars/#/composeconvertwindowspaths. At any rate, you have to figure out what the volume call you're making will be expanded to.

Note that this only applies in Docker Toolbox/Machine land. I'm not sure about D4W

**Comment by matheuscmpm** [comment]
@cyli Thanks for you help. I was able to get rid of my problem. The problem was that my entrypoint was not being correctly executed. On mac/unix it was fine, but in Windows I was getting errors which I believed that was caused by volumes not being mounted correctly. I converted my entrypoint scripts using this command in my Dockerfile: 

RUN sed -i -e 's/\r$//' /home/init.sh

Now "everything" is mounting right. My only problem is that this script uses some database integration to run scripts to version my database, like migrations. But in Windows, my application cant find the database connection, only If I exec this command after getting the container executed. But I guess this isn't a docker problem. 

Thanks again for your time

**Comment by DanielAron** [comment]
Same issue here at Windows 7

```
docker --version
time="2017-02-21T16:21:14-03:00" level=info msg="Unable to use system certificate pool: crypto/x509: system root pool is not available on Windows"
Docker version 1.13.1, build 092cba3
```

**Comment by cyli** [comment]
@matheuscmpm Cool, thanks for checking back in!  Glad that particular issue is fixed.  :)

Thanks @DanielAron - https://github.com/docker/go-connections/issues/30 is an attempt to fix this for docker-machine hosts.

**Comment by archae0pteryx** [comment]
I'm experiencing the same thing. I have read the thread, and it sounds like ya'll know what the problem is. Do you guys need my formal "issue" submission? Sounds like you've got your leg over the ol' girl by now though. I'll check back and see if this has been resolved soon. Cheers ya'll.

**Comment by wenquanzhao** [comment]
@nerro  how to make server-experimental true?

**Comment by thaJeztah** [comment]
@wenquanzhao see  https://github.com/docker/docker/tree/v1.13.1/experimental#use-docker-experimental, but please don't hijack issues with unrelated questions

**Comment by rainabba** [comment]
A month later, the first thing shown after install is "Docker is now up and running!" and an animation indicating we should run `docker info`, `docker version` and such, then this. Just downloaded from https://docs.docker.com/docker-for-windows/install/

I did grab the beta (because I need to be able to switch between Windows and CentOS containers and something led me to think beta was still needed for that), but 30 days later and even the version command can't be run?

Since I hoped the issue was specific to the beta, I proceeded with uninstalling to install the stable (hoping I can still run both Windows and CentOS containers) only to find out that the bug isn't limited to beta.

Not very confidence-inspiring, especially when I've been watching and wondering if docker was production ready since Nov 2015 :/ 

Can anyone provide an update on this? Is there a nightly build with a fix or such?

**Comment by lmakarov** [comment]
The milestone on this issue should be updated to `17.04.0` to match the one on the PR that is supposed to fix it: https://github.com/docker/docker/pull/30729

**Comment by thaJeztah** [comment]
@lmakarov there was a second pull request against the 1.13.x branch (and part of 17.03) https://github.com/docker/docker/pull/30730. Let me update the other PR to prevent confusion

**Comment by lmakarov** [comment]
@thaJeztah I head to re-read the whole thread to get the answer as I can still see the issue on a Windows host when used with a docker-machine provisioned VM.

This answered my question
https://github.com/docker/docker/issues/30450#issuecomment-281498295

I'm assuming this PR https://github.com/docker/go-connections/pull/30 should be part of the `17.04.0` milestone?

**Comment by cyli** [comment]
@lmakarov Yes, plus a docker change to actually use the flag.  Am working on that vendor and flag as a PR now.

**Comment by feryardiant** [comment]
Me also getting same issue since version 1.13.x until now 17.03.0-ce.
![image](https://cloud.githubusercontent.com/assets/508665/24129186/67d1da64-0e13-11e7-87da-4724fc1d087a.png)

**Comment by ranneyd** [comment]
@feryardiant +1

@lmakarov so was this added to 17.04 as far as you can tell? And when do we think 17.04 will go out? Thanks!

**Comment by feryardiant** [comment]
@raesene unfortunately I can't see it's working since installed few month ago when I started move to Windows also I'm still learn it, so I can't tell you anything. Sorry.

**Comment by sebdelvalle** [comment]
+1
a setting to hide this would be enough i think

**Comment by laurencefass** [comment]
ive got the same problem and docker attach and docker exec both raise the message and leave the command line hanging.

**Comment by sdesalas** [comment]
Still happening on latest version downloaded via website:

```
$ docker --version
time="2017-04-08T17:13:48+10:00" level=info msg="Unable to use system certificate pool: crypto/x509: system root pool is not available on Windows"
Docker version 17.03.1-ce, build c6d412e
```

**Comment by cyli** [comment]
The fix will be in 17.04, which is in rc's now.

**Comment by seriousme** [comment]
hate to spoil the party, but I just upgraded to 17.04 and still have the issue
```
 docker version
time="2017-04-08T18:41:03+02:00" level=info msg="Unable to use system certificate pool: crypto/x509: system root pool is not available on Windows"
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 08:47:51 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Wed Apr  5 18:45:47 2017
 OS/Arch:      linux/amd64
 Experimental: false
```
